### PR TITLE
add support for MT7981

### PR DESCRIPTION
**Pull request by dangowrt**
@lynxis had prepared patches for MT7981 WiFi and recently @csyuanc has posted firmware files

Let's discuss details here

### CODE REVIEW

**Comment by zhaojh329** [comment]
@dangowrt 
Is it well to rename CONFIG_MT7986_WMAC to CONFIG_MT798X_WMAC?

**Comment by ptpt52** [comment]
mark

**Comment by zhaojh329** [comment]
I tested in GL-MT3000, it's only about 800Mbps.
It's about 1.7Gbps with sdk driver.

**Comment by dangowrt** [comment]
To test this in OpenWrt:
```diff
From bc7e0f230467bee84b4da135cc2518b6be3dfc24 Mon Sep 17 00:00:00 2001
From: Daniel Golle <daniel@makrotopia.org>
Date: Tue, 18 Apr 2023 23:52:38 +0100
Subject: [PATCH] mt76: add support for MT7981

---
 package/kernel/mt76/Makefile | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/package/kernel/mt76/Makefile b/package/kernel/mt76/Makefile
index 34930d7693..f180a7c04c 100644
--- a/package/kernel/mt76/Makefile
+++ b/package/kernel/mt76/Makefile
@@ -6,11 +6,11 @@ PKG_RELEASE=5
 PKG_LICENSE:=GPLv2
 PKG_LICENSE_FILES:=
 
-PKG_SOURCE_URL:=https://github.com/openwrt/mt76
+PKG_SOURCE_URL:=https://github.com/dangowrt/mt76
 PKG_SOURCE_PROTO:=git
-PKG_SOURCE_DATE:=2023-03-01
-PKG_SOURCE_VERSION:=c32d6d849c43792abd8007e13e468b12d6d6e0b7
-PKG_MIRROR_HASH:=b7004bc920ba44cef2f7868d94beb2d288ff9d399af624ce5dad972f953723c8
+PKG_SOURCE_DATE:=2023-04-18
+PKG_SOURCE_VERSION:=075e393e01eef0800825ecb04de66ae003aeb373
+PKG_MIRROR_HASH:=b4869c7135a438a77fc7ebf1616fd030abdefceaadb9a26556c6906eb1ec593b
 
 PKG_MAINTAINER:=Felix Fietkau <nbd@nbd.name>
 PKG_USE_NINJA:=0
@@ -245,6 +245,12 @@ define KernelPackage/mt7916-firmware
   TITLE:=MediaTek MT7916 firmware
 endef
 
+define KernelPackage/mt7981-firmware
+  $(KernelPackage/mt76-default)
+  DEPENDS:=@TARGET_mediatek_filogic
+  TITLE:=MediaTek MT7981 firmware
+endef
+
 define KernelPackage/mt7986-firmware
   $(KernelPackage/mt76-default)
   DEPENDS:=@TARGET_mediatek_filogic
@@ -510,6 +516,15 @@ define KernelPackage/mt7986-firmware/install
 		$(1)/lib/firmware/mediatek
 endef
 
+define KernelPackage/mt7981-firmware/install
+	$(INSTALL_DIR) $(1)/lib/firmware/mediatek
+	cp \
+		$(PKG_BUILD_DIR)/firmware/mt7981_wa.bin \
+		$(PKG_BUILD_DIR)/firmware/mt7981_wm.bin \
+		$(PKG_BUILD_DIR)/firmware/mt7981_rom_patch.bin \
+		$(1)/lib/firmware/mediatek
+endef
+
 define KernelPackage/mt7921-firmware/install
 	$(INSTALL_DIR) $(1)/lib/firmware/mediatek
 	cp \
@@ -548,6 +563,7 @@ $(eval $(call KernelPackage,mt7663s))
 $(eval $(call KernelPackage,mt7915-firmware))
 $(eval $(call KernelPackage,mt7915e))
 $(eval $(call KernelPackage,mt7916-firmware))
+$(eval $(call KernelPackage,mt7981-firmware))
 $(eval $(call KernelPackage,mt7986-firmware))
 $(eval $(call KernelPackage,mt7921-firmware))
 $(eval $(call KernelPackage,mt7921-common))
-- 
2.40.0
```

**Comment by dangowrt** [comment]
This is the patch to load the correct WO firmware:
https://patchwork.kernel.org/project/netdevbpf/patch/2651cfd1427175ac1ad08d5b7be146c55d94b674.1681930898.git.daniel@makrotopia.org/

**Comment by dangowrt** [comment]
Measuring performance with iperf3 using `bridger` to enable WED on bridged 2.5G Ethernet on HE160 channel:
```
[daniel@mixxxtop ~]$ iperf3 -c 192.168.1.2
Connecting to host 192.168.1.2, port 5201
[  5] local 192.168.1.143 port 40534 connected to 192.168.1.2 port 5201
[ ID] Interval           Transfer     Bitrate         Retr  Cwnd
[  5]   0.00-1.00   sec  62.8 MBytes   527 Mbits/sec    3    339 KBytes       
[  5]   1.00-2.00   sec   121 MBytes  1.01 Gbits/sec    0    551 KBytes       
[  5]   2.00-3.00   sec   137 MBytes  1.15 Gbits/sec    0    720 KBytes       
[  5]   3.00-4.00   sec   141 MBytes  1.18 Gbits/sec    3    848 KBytes       
[  5]   4.00-5.00   sec   145 MBytes  1.22 Gbits/sec    0    973 KBytes       
[  5]   5.00-6.00   sec   156 MBytes  1.31 Gbits/sec    0   1.07 MBytes       
[  5]   6.00-7.00   sec   161 MBytes  1.35 Gbits/sec    0   1.17 MBytes       
[  5]   7.00-8.00   sec   161 MBytes  1.35 Gbits/sec    4   1.26 MBytes       
[  5]   8.00-9.00   sec   171 MBytes  1.44 Gbits/sec    0   1.36 MBytes       
[  5]   9.00-10.00  sec   168 MBytes  1.41 Gbits/sec    0   1.45 MBytes       
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.00  sec  1.39 GBytes  1.20 Gbits/sec   10             sender
[  5]   0.00-10.01  sec  1.39 GBytes  1.19 Gbits/sec                  receiver

iperf Done.
[daniel@mixxxtop ~]$ iperf3 -R -c 192.168.1.2
Connecting to host 192.168.1.2, port 5201
Reverse mode, remote host 192.168.1.2 is sending
[  5] local 192.168.1.143 port 59256 connected to 192.168.1.2 port 5201
[ ID] Interval           Transfer     Bitrate
[  5]   0.00-1.00   sec   148 MBytes  1.24 Gbits/sec                  
[  5]   1.00-2.00   sec   177 MBytes  1.49 Gbits/sec                  
[  5]   2.00-3.00   sec   179 MBytes  1.50 Gbits/sec                  
[  5]   3.00-4.00   sec   183 MBytes  1.53 Gbits/sec                  
[  5]   4.00-5.00   sec   182 MBytes  1.52 Gbits/sec                  
[  5]   5.00-6.00   sec   179 MBytes  1.50 Gbits/sec                  
[  5]   6.00-7.00   sec   182 MBytes  1.53 Gbits/sec                  
[  5]   7.00-8.00   sec   185 MBytes  1.55 Gbits/sec                  
[  5]   8.00-9.00   sec   184 MBytes  1.55 Gbits/sec                  
[  5]   9.00-10.00  sec   176 MBytes  1.48 Gbits/sec                  
- - - - - - - - - - - - - - - - - - - - - - - - -
[ ID] Interval           Transfer     Bitrate         Retr
[  5]   0.00-10.01  sec  1.74 GBytes  1.49 Gbits/sec  146             sender
[  5]   0.00-10.00  sec  1.73 GBytes  1.49 Gbits/sec                  receiver

iperf Done.
```

**Comment by dangowrt** [comment]
Complete OpenWrt tree for the GL.iNet GL-MT3000, incl. this PR as well as patch to load correct WO firmware, but not the WO firmware itself. To use WO, place `mt7981_wo.bin` in `files/lib/firmware/mediatek` folder.

https://github.com/dangowrt/openwrt/tree/gl-mt3000

**Comment by nicefile** [comment]
So I've try my best to enable wifi on Cudy WR3000 .Basically I'm using all patches from branch https://github.com/dangowrt/openwrt/tree/gl-mt3000 and wifi stay undetected 
only thing I've got visible when I've changed

**Comment by nicefile** [comment]
> > `reg = <0x00 0x18000000 0x00 0x1000000>;`
> 
> You should not need to change `mt7981.dtsi`.
> 
> We are looking at https://github.com/nicefile/openwrt-vector/tree/wr3000 , right?
> 
> Did you select `kmod-mt7981-firmware`? I saw it was missing in [nicefile/openwrt-vector@c21abdb#r109817109](https://github.com/nicefile/openwrt-vector/commit/c21abdb6c36b239df1e6b57b9ad372287043be28#r109817109)

Yes. I have kmod-mt7981-firmware and wo.bin in place . mt7615e doesn't initiate wireless like there is some missing GPIO or reg value . I've looked in OEM dts and its firmware for any trace of it but no luck

**Comment by JiaY-shi** [comment]
> 

@dangowrt The performance test results I did on gl-mt3000 using iperf3:
```
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-60.00  sec   853 MBytes   119 Mbits/sec                  sender
[  4]   0.00-60.00  sec   853 MBytes   119 Mbits/sec                  receiver
[  6]   0.00-60.00  sec   836 MBytes   117 Mbits/sec                  sender
[  6]   0.00-60.00  sec   836 MBytes   117 Mbits/sec                  receiver
[  8]   0.00-60.00  sec   821 MBytes   115 Mbits/sec                  sender
[  8]   0.00-60.00  sec   821 MBytes   115 Mbits/sec                  receiver
[ 10]   0.00-60.00  sec   809 MBytes   113 Mbits/sec                  sender
[ 10]   0.00-60.00  sec   809 MBytes   113 Mbits/sec                  receiver
[ 12]   0.00-60.00  sec   796 MBytes   111 Mbits/sec                  sender
[ 12]   0.00-60.00  sec   796 MBytes   111 Mbits/sec                  receiver
[ 14]   0.00-60.00  sec   769 MBytes   107 Mbits/sec                  sender
[ 14]   0.00-60.00  sec   769 MBytes   107 Mbits/sec                  receiver
[ 16]   0.00-60.00  sec   738 MBytes   103 Mbits/sec                  sender
[ 16]   0.00-60.00  sec   738 MBytes   103 Mbits/sec                  receiver
[ 18]   0.00-60.00  sec   720 MBytes   101 Mbits/sec                  sender
[ 18]   0.00-60.00  sec   720 MBytes   101 Mbits/sec                  receiver
[SUM]   0.00-60.00  sec  6.19 GBytes   887 Mbits/sec                  sender
[SUM]   0.00-60.00  sec  6.19 GBytes   887 Mbits/sec                  receiver
```
I am using HE80 channel, when I switch to HE160 channel, the wifi is lost.
PS: The other side is Intel AX211

**Comment by mrkiko** [comment]
Mark.

**Comment by dangowrt** [comment]
> @dangowrt The performance test results I did on gl-mt3000 using iperf3:

Are you running `iperf3` on the device itself? If so, expect way lower performance than the actual routing/forwarding/bridge performance as `iperf3` itself will consume most of the available CPU cycles...

My test above is done running iperf3 server on a powerful host connected to the 2.5G interface of the MT3000 and i7 based laptop with Intel AX210 connecting via WiFi runs the iperf3 client.

**Comment by JiaY-shi** [comment]
> > @dangowrt The performance test results I did on gl-mt3000 using iperf3:
> 
> Are you running on the device itself? If so, expect way lower performance than the actual routing/forwarding/bridge performance as itself will consume most of the available CPU cycles...`iperf3``iperf3`
> 
> My test above is done running iperf3 server on a powerful host connected to the 2.5G interface of the MT3000 and i7 based laptop with Intel AX210 connecting via WiFi runs the iperf3 client.

@dangowrt Thanks for your reply.
Indeed as you said, I run iperf3 on gl-mt3000. Limited by the bandwidth of the LAN, only Gigabit, so I can't use your way to test.
PS: Why can't I use the HE160 channel? When I switch to HE160, the wifi is lost, do you know why?

**Comment by dangowrt** [comment]
> PS: Why can't I use the HE160 channel? When I switch to HE160, the wifi is lost, do you know why?

First of all 160 MHz channels afaik only work if you set the country code, because there isn't 160 MHz continuous spectrum without using DFS channels. All ETSI/EU and FCC countries should allow you to use HE160 if you set channel to 'auto'.

**Comment by JiaY-shi** [comment]
> > PS: Why can't I use the HE160 channel? When I switch to HE160, the wifi is lost, do you know why?
> 
> First of all 160 MHz channels afaik only work if you set the country code, because there isn't 160 MHz continuous spectrum without using DFS channels. All ETSI/EU and FCC countries should allow you to use HE160 if you set channel to 'auto'.

@dangowrt Thanks for your reply, I forgot to adjust the region before, now I use the correct region, and HE160 can be used normally, here is the new performance test result:
```
[ ID] Interval           Transfer     Bandwidth
[  4]   0.00-60.00  sec  1.66 GBytes   237 Mbits/sec                  sender
[  4]   0.00-60.00  sec  1.66 GBytes   237 Mbits/sec                  receiver
[  6]   0.00-60.00  sec  1.62 GBytes   232 Mbits/sec                  sender
[  6]   0.00-60.00  sec  1.62 GBytes   232 Mbits/sec                  receiver
[  8]   0.00-60.00  sec  1.58 GBytes   227 Mbits/sec                  sender
[  8]   0.00-60.00  sec  1.58 GBytes   227 Mbits/sec                  receiver
[ 10]   0.00-60.00  sec  1.54 GBytes   221 Mbits/sec                  sender
[ 10]   0.00-60.00  sec  1.54 GBytes   221 Mbits/sec                  receiver
[ 12]   0.00-60.00  sec  1.50 GBytes   215 Mbits/sec                  sender
[ 12]   0.00-60.00  sec  1.50 GBytes   215 Mbits/sec                  receiver
[ 14]   0.00-60.00  sec  1.47 GBytes   211 Mbits/sec                  sender
[ 14]   0.00-60.00  sec  1.47 GBytes   211 Mbits/sec                  receiver
[ 16]   0.00-60.00  sec  1.43 GBytes   205 Mbits/sec                  sender
[ 16]   0.00-60.00  sec  1.43 GBytes   205 Mbits/sec                  receiver
[ 18]   0.00-60.00  sec  1.37 GBytes   196 Mbits/sec                  sender
[ 18]   0.00-60.00  sec  1.37 GBytes   196 Mbits/sec                  receiver
[SUM]   0.00-60.00  sec  12.2 GBytes  1.74 Gbits/sec                  sender
[SUM]   0.00-60.00  sec  12.2 GBytes  1.74 Gbits/sec                  receiver
```
GL-MT3000 acts as the server, and Intel AX211 acts as the client.
Thank you for your excellent work.

**Comment by nicefile** [comment]
so look look like watchdog declaration is mandatory  .After adding this to my DTS . Wireless start working
`&watchdog {
	status = "okay";
};`

**Comment by dangowrt** [comment]
> so look look like watchdog declaration is mandatory

Yes. watchdog also serves as a reset controller and is used to reset wireless parts.

**Review by nicefile — APPROVED** [review]

**Review by nicefile — APPROVED** [review]
tested on Cudy WR3000 and found no issues. WET mode works if bridger does its job.

File: mt7915/soc.c [hunk 1]
```diff
@@ -744,12 +784,19 @@ mt7986_wmac_afe_cal(struct mt7915_dev *dev, u8 adie, bool dbdc, u32 adie_type)
 		       MT_AFE_RG_WBG_EN_WPLL_UP_MASK, 0x1);
 	usleep_range(60, 100);
 
-	mt76_rmw_field(dev, MT_AFE_DIG_EN_01(idx),
-		       MT_AFE_RG_WBG_EN_TXCAL_MASK, 0x1f);
+	/* TODO: mt7981: sets also bit WF4, but mt7986 doesn't need/allow this? */
```
Comment (dangowrt) [inline mt7915/soc.c:787]: @LorenzoBianconi @nbd168 any idea why WF4 should not be set for MT7986?

**Comment by 1248289414** [comment]
After testing on the Xiaomi [WR30U](https://github.com/1248289414/openwrt/commit/9d680c58bd7e39456da527378eaff5f22aab3d5e), `ap` mode works no problem. But `sta` mode works a bit weird: the device can work in `ac+160Mhz` mode to get 1560.0 Mbit/s and in `ax+20Mhz` to get 286.7 MBit/s, but `ax+40/80/160Mhz` only gets 6.0 MBit/s, the connection is slow or unavailable.

Do you have any suggestions about this?

The AP device is TP-XDR3050(IPQ0509 + QCN6102).

**Comment by dangowrt** [comment]
@1248289414 It'd be very interesting to know if this problem also affects MT7986. Do you happen to have a device with MT7986 for testing?

File: mt7915/soc.c [hunk 2]
```diff
@@ -489,7 +519,10 @@ static int mt7986_wmac_adie_patch_7976(struct mt7915_dev *dev, u8 adie)
 		rg_xo_01 = 0x1d59080f;
 		rg_xo_03 = 0x34c00fe0;
 	} else {
-		rg_xo_01 = 0x1959f80f;
+		if (is_mt7981(&dev->mt76))
+			rg_xo_01 = 0x1959f80f;
```
Comment (lynxis) [inline mt7915/soc.c:523]: This must be 0x1959c80f for mt7981

**Review by lynxis — CHANGES_REQUESTED** [review]

File: mt7915/soc.c (hunk 2 above)
Comment (dangowrt) [reply mt7915/soc.c]: I've re-picked the commit from your tree and also slightly improved the commit message.

**Comment by nicefile** [comment]
@dangowrt @1248289414 i can confirm that as well.
briefly tested 5Ghz AC /AX to mt7621+mt7915 OpenWrt router (AX80/Ch36) . For me even AC and AX20 was unstable .

**Comment by dangowrt** [comment]
@nicefile Did you test client mode after https://github.com/openwrt/mt76/pull/771/commits/933a1815d19d7b2cf1d5592558e8132813537944 was pushed?

**Comment by V10lator** [comment]
Tested this in a CUDY WR3000. Some findings:
- MAC adresses are wrong. Like with original FW both where from CUDY but now the one on 2.4 GHz is from ralink and the one on the 5 GHz radio is unknown.
- Scanning (LUCI Channel Analysis) disconnects clients.

**Comment by dangowrt** [comment]
> Tested this in a CUDY WR3000. Some findings:
> 
>     * MAC adresses are wrong

MAC addresses are device-specific, please report this to @nicefile or comment here: https://github.com/nicefile/openwrt-vector/tree/wr3000

>     * Scanning (LUCI Channel Analysis) disconnects clients.

Unfortunately this is the expected behavior as there is no dedicated scanning radio or antenna in this device (and also not on the MT3000).

**Review by lynxis — APPROVED** [review]

**Comment by lynxis** [comment]
@dangowrt do we want to rename the symbol to CONFIG_MT798X_WMAC as zhaojh329 recommended? I've squashed it into a commit [here](https://github.com/lynxis/mt76/commit/12544cc7de4de0c1dfa43c89e7a5d0bf69795e99)

Pull it if you like.

**Comment by dangowrt** [comment]
@lynxis I've squashed that change, though I'm not sure if upstream will like it, I remember that some maintainers don't like using wildcard names but that mostly applies to device tree and for a Kconfig symbol it may be ok.

**Comment by V10lator** [comment]
Looks like one of the last force-pushes broke anything: Driver not working anymore. Nothing in the syslogs.

**Review by ptpt52 — APPROVED** [review]

**Comment by ptpt52** [comment]
everthing works fine for me.

**Comment by JiaY-shi** [comment]
> Looks like one of the last force-pushes broke anything: Driver not working anymore. Nothing in the syslogs.

Did you not modify the file `package/kernel/mt76/Makefile` in openwrt source?
Like this:
```
diff --git a/package/kernel/mt76/Makefile b/package/kernel/mt76/Makefile
index 34930d7693..cc574fa27e 100644
--- a/package/kernel/mt76/Makefile
+++ b/package/kernel/mt76/Makefile
@@ -374,8 +374,8 @@ endif
 ifdef CONFIG_PACKAGE_kmod-mt7915e
   PKG_MAKE_FLAGS += CONFIG_MT7915E=m
   ifdef CONFIG_TARGET_mediatek_filogic
-    PKG_MAKE_FLAGS += CONFIG_MT7986_WMAC=y
-    NOSTDINC_FLAGS += -DCONFIG_MT7986_WMAC
+    PKG_MAKE_FLAGS += CONFIG_MT798X_WMAC=y
+    NOSTDINC_FLAGS += -DCONFIG_MT798X_WMAC
   endif
 endif
 ifdef CONFIG_PACKAGE_kmod-mt7921-common
```

**Comment by V10lator** [comment]
@JiaY-shi Ofc. this was the issue. Sorry for the trouble.

**Comment by nbd168** [comment]
This kind of development should not happen in GitHub PRs and merged from there. Please send these changes to linux-wireless.

**Comment by dangowrt** [comment]
@nbd168 Done.
https://marc.info/?l=linux-wireless&m=168398543818756
https://patchwork.kernel.org/project/linux-wireless/patch/ZF-SN-sElZB_g_bA@pidgin.makrotopia.org/

**Comment by nbd168** [comment]
thanks!

**Pull request merged**